pkg/controllermanager/controller/managedseedset/managedseedset_controller.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedseedset
+
+import (
+	"k8s.io/client-go/tools/cache"
+
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+)
+
+func (c *Controller) managedSeedSetAdd(obj interface{}) {
+	_, ok := obj.(*seedmanagementv1alpha1.ManagedSeedSet)
+	if !ok {
+		return
+	}
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+
+	c.managedSeedSetQueue.Add(key)
+}
+
+func (c *Controller) managedSeedSetUpdate(_, newObj interface{}) {
+	set, ok := newObj.(*seedmanagementv1alpha1.ManagedSeedSet)
+	if !ok {
+		return
+	}
+	if set.Generation == set.Status.ObservedGeneration {
+		return
+	}
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		return
+	}
+
+	c.managedSeedSetQueue.Add(key)
```

Comment:
I observed multiple cases (with the `CachedRuntimeClients` feature gate enabled), where the finalizer addition led to a immediate requeue and second reconciliation of the set.
In such cases the shoot cache wasn't synced fast enough so that the controller didn't recognize that it already created some replica and thus immediately created another one (although `spec.replicas` was set to `1`).

There are probably multiple ways to solve this. Some ideas would be:
- don't enqueue if we detect that the finalizer was added (not really correct)
- update the observed generation even before adding the finalizer (would also not solve the root cause with a stale cache)
- fetch replicas using an API reader (also not quite nice, because of the network traffic)
- fetch a metadata only list of replicas using an API reader and cross check the count with the list from the cache

Anyways, I'm not quite sure yet, what's the proper way to solve this.
Maybe you have a good idea? Maybe we can chat about it or even look at some k8s code, how they do it.

Reply:
It seems to me only your fourth proposal is kind of good enough. I don't have a better idea.

I wonder though, what actually happened next? OK, an extra replica was created, but at some point it was deleted, to reach the target count, correct? Perhaps we could live with this, at least for now?

Reply:
The interesting part is that I am also testing with `CachedRuntimeClients` feature gate enabled (and have been doing that all the time), but never observed anything like you describe. Did you do something in addition, or is there a way to reliably reproduce the above issue?

Reply:
I was not able to reproduce this, but implemented the "fourth proposal" anyway.